### PR TITLE
fix(animations): ensure all child elements are rendered before running animations

### LINKED ISSUE #9402: RC2 does not animate nested elements correctly when using AUTO STYLE feature
**I'm submitting a problem with animations of nested elements in RC2**
```
[x] bug report
[ ] feature request
```
! First of all thanks a lot for AUTO STYLE. It is just an excellent feature that brings great opportunities and save a lot of code and time !

**Current behavior**
Plunker demo: There is an outer DIV that contains another inner DIV. Buttons allow to expand and collapse them individually.
- Expand the outer and then inner.
- Continue with collapsing the outer.
- Finally expand the outer again.

It is possible to see that height animation using auto style "*" works correctly only partly - it does not respect the height of previously expanded content of inner DIV. It looks like that the hidden evaluation of the outer DIV height does not see the inner expanded DIV. Probably something is wrong with evaluation of the template where *ngIf directives are used.

**Expected/desired behavior**
Animation of the outer DIV should respect its full content which is available in the actual moment.

**Reproduction of the problem**
(Example: http://plnkr.co/edit/QXiq9bYcCAcRRmxx04BS?p=preview)

**Please tell us about your environment:**
* **Angular version:** 2.0.0-RC.2
* **Browser:** [Chrome] 
* **Language:** [TypeScript]

### PR DESCRIPTION
**Pull request by matsko**
Closes #9402
Closes #9775

### CODE REVIEW

**Comment by matsko** [comment]
This is succeeded from https://github.com/angular/angular/pull/9775

File: modules/@angular/compiler/src/view_compiler/property_binder.ts [hunk 1]
```diff
@@ -8,6 +8,7 @@
 
 import {EMPTY_STATE as EMPTY_ANIMATION_STATE, LifecycleHooks, isDefaultChangeDetectionStrategy} from '../../core_private';
 import * as cdAst from '../expression_parser/ast';
+import {Map} from '../facade/collection';
```
Comment (vicb) [inline modules/@angular/compiler/src/view_compiler/property_binder.ts]: not needed

File: modules/@angular/compiler/src/view_compiler/property_binder.ts [hunk 2]
```diff
@@ -36,6 +37,8 @@ function createCurrValueExpr(exprIndex: number): o.ReadVarExpr {
   return o.variable(`currVal_${exprIndex}`);  // fix syntax highlighting: `
 }
 
+var _animationViewCheckedFlagMap = new Map<CompileView, boolean>();
```
Comment (vicb) [inline modules/@angular/compiler/src/view_compiler/property_binder.ts]: const

File: modules/@angular/compiler/src/view_compiler/property_binder.ts [hunk 3]
```diff
@@ -171,6 +174,12 @@ function bindAndWriteToRenderer(
             animation.fnVariable.callFn([o.THIS_EXPR, renderNode, oldRenderValue, emptyStateValue])
                 .toStmt());
 
+        if (!_animationViewCheckedFlagMap.get(view)) {
+          _animationViewCheckedFlagMap.set(view, true);
```
Comment (vicb) [inline modules/@angular/compiler/src/view_compiler/property_binder.ts]: add test to check if the flag is set to false

File: modules/@angular/core/test/animation/animation_integration_spec.ts [hunk 4]
```diff
@@ -26,8 +30,7 @@ export function main() {
     declareTests({useJit: false});
   } else {
     describe('jit', () => { declareTests({useJit: true}); });
-
-    describe('no jit', () => { declareTests({useJit: false}); });
+    // describe('no jit', () => { declareTests({useJit: false}); });
```
Comment (vicb) [inline modules/@angular/core/test/animation/animation_integration_spec.ts]: debug

File: modules/@angular/core/testing/testing_internal.ts [hunk 5]
```diff
@@ -13,7 +13,7 @@ import {Math, global, isFunction, isPromise} from '../src/facade/lang';
 import {AsyncTestCompleter} from './async_test_completer';
 import {getTestInjector, inject} from './test_injector';
 
-export {MockAnimationPlayer} from './animation/mock_animation_player';
+export {MockAnimationPlayer} from '../../platform-browser/testing/mock_animation_player';
```
Comment (vicb) [inline modules/@angular/core/testing/testing_internal.ts]: `@angular/...`

File: modules/@angular/platform-browser/src/dom/web_animations_player.ts [hunk 6]
```diff
@@ -33,13 +42,46 @@ export class WebAnimationsPlayer implements AnimationPlayer {
     }
   }
 
+  init(): void {
+    if (this._initialized) return;
+    this._initialized = true;
+
+    var anyElm = <any>this.element;
```
Comment (vicb) [inline modules/@angular/platform-browser/src/dom/web_animations_player.ts]: not needed

**Pull request closed without merging**